Add additional tests for blackboard types PR

bjoernschroeder opened the pull request.

bjoernschroeder committed:
add bb tests for up and down conversion

bjoernschroeder committed:
add UnknownType class to tests, add tests for get exceptions

bjoernschroeder committed:
add tests for setting and initialization, add non constructable helper class, rename test fixture for blackboard

bjoernschroeder committed:
add set and init test for blackboard

bjoernschroeder committed:
use EXPECT_THROW for exception testing

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -117,5 +129,638 @@ TEST_F(TestBlackBoard, testMapping)
+    auto srcLocked = LockedBlackboardRW(srcBb);
+    auto targetLocked = LockedBlackboardRW(targetBb);
+
+    srcLocked.set<int64_t>("valueSrc", 1);
```
I think you can templatize the mapping checks & put them all under one (or a few related) test(s)
```
template <class srcType, class targetType>
bool checkMapping(srcBB, targetBB, srcValue, targetValue, mappedValue);
```

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -117,5 +129,638 @@ TEST_F(TestBlackBoard, testMapping)
+                try {
+                    bb_locked.get<UnknownType>("value");
+                } catch (const BlackboardException& e) {
+                    EXPECT_STREQ(e.what(), "Blackboard exception: get() type mismatch, key: value, setType: unknown, getType: unknown");
```
I think its more consistent to have setType as std::any here, so essentially setType refects the varaint's type. getType can remain as it is. So, just have to change the exception string that is thrown

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -117,5 +129,638 @@ TEST_F(TestBlackBoard, testMapping)
+TEST_F(TestBlackboard, testSetNotMatchingKnownType)
+{
+    std::unique_ptr<alica::BlackboardBlueprint> blueprint = std::make_unique<alica::BlackboardBlueprint>();
+    blueprint->addKey("intVal", "int64", "14");
```
perhaps we can define an enum for the type name indexes & use BB_VALUE_TYPE_NAMES to access the type names instead of hard coded string literals i.e.
```
enum { BOOL=0, INT_64=1....}
BB_VALUE_TYPE_NAMES[BOOL], BB_VALUE_TYPE_NAMES[INT_64]
```
or just string constants 
```
constexpr const char* BOOL = "bool"
```

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -117,5 +129,638 @@ TEST_F(TestBlackBoard, testMapping)
+    EXPECT_EQ(bb_locked.get<PlanStatus>("value"), PlanStatus::Success);
+
+    bb_locked.set<UnknownType>("value", UnknownType());
+    EXPECT_EQ(typeid(UnknownType), typeid(bb_locked.get<UnknownType>("value")));
```
types are definitely going to match because get returns the type that it was called with. What needs to be checked is if the values match. You can make unknown type a wrapper for an int i.e. add a constructor which takes an int & check if the right value is returned

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -117,5 +129,638 @@ TEST_F(TestBlackBoard, testMapping)
+                    alica::Blackboard(blueprint.get());
+                } catch (const BlackboardException& e) {
+                    EXPECT_STREQ(e.what(), "Blackboard exception: Could not set key: anyWithDefault, from default value: defaultValue, type: int64, details: "
+                                           "yaml-cpp: error at line 1, column 1: bad conversion");
```
this can become flaky really soon. Can't rely on the exception msg thrown by yaml cpp library. Best not to check the excpetion msg. Should suffice to just check if an exception is thrown

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -117,5 +129,638 @@ TEST_F(TestBlackBoard, testMapping)
+    alica::Blackboard blackboard = alica::Blackboard(blueprint.get());
+    alica::LockedBlackboardRO bb = LockedBlackboardRO(blackboard);
+
+    EXPECT_EQ(bb.get<int64_t>("intVal"), 0);
```
best to use the default construtor i.e int64_t(), double() etc

veerajsk reviewed (commented):
Some scenarios not covered by tests especially some tricky ones:
- yaml type is std::any. This has a bunch of interesting possibilities & tricky cases
      - set the key with known, unknown type & std::any. get with right type or use invalid type
      - map from a key with has known, unknown or std::any yaml type
- yaml type is a known type & try to set with a type that is not the same but convertible to that type eg. yaml type is int64_t, but set type is int32_t, int8_t & possibly unsigned types (actually I'm not sure how unsigned types will be handled especially uint64_t. Perhaps we need to add a uint64_t type to the variant)
- call set without specifying the type i.e. `set("key", value)`. Then type will be deduced.

bjoernschroeder committed:
- use template for mapping tests
- use constexpr for type names
- improve exception tests, dont check exception msg
- compare UnknownTypes by using an int value, passed during initialization
- add additional tests for std::any

bjoernschroeder committed:
add tests setWithConvertibleType and setWithoutSpecifyingType, use default constructor to create default values of base types

bjoernschroeder committed:
activate tests for different types convertible to int64_t

bjoernschroeder committed:
format

bjoernschroeder committed:
remove old code

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -117,5 +145,452 @@ TEST_F(TestBlackBoard, testMapping)
+    targetLocked.set<int64_t>("valueTarget", 0);
+    EXPECT_TRUE((checkMapping<int64_t>(srcBb.impl(), targetBb.impl(), "valueSrc", "valueTarget", -1)));
+
+    targetLocked.set<double>("valueTarget", 0.0);
```
as I had said map is an internal method used to map keys that are set in the pml file. You have instead used set to influence the type of the target. That is relying on internal behaviour. Anyway I'll fix it

veerajsk committed:
Set yaml type for blackboard mapping tests
